[#39] Move to Ansifmt

chshersh opened the pull request:
Closes #39 

It was surprisingly easy to migrate! I haven't noticed any regressions in the behaviour. Label background and foreground based on luminance still work as expected:

![image](https://github.com/user-attachments/assets/84338d52-3cb5-4f4a-9228-8966792e6d1b)

I'm glad that quite a lot of bespoke logic got deleted, and there's a smaller context to keep in mind while working on this project.

![image](https://github.com/user-attachments/assets/44ee9c78-2a4b-4d32-bae7-0815b6dab017)

Have a look and let me know if it makes sense!

chshersh committed:
Move to Ansifmt

chshersh commented on github_tui.opam.template:
```diff
@@ -1,3 +1,4 @@
 pin-depends: [
   ["shape-the-term.dev" "git+https://github.com/jmcavanillas/shape-the-term"]
+  ["ansifmt.dev" "git+https://github.com/qexat/ansifmt"]
```
Depending on a GitHub version until it's merged to opam-repository:

- https://github.com/ocaml/opam-repository/pull/27357

qexat replied:
v0.2.0 was just released, so I updated that

chshersh commented on lib/pretty/color.ml:
Glad this is gone! 🚀

chshersh commented on lib/pretty/pretty.mli:
```diff
@@ -1,5 +1,8 @@
 (** A library for declarative UI definition. *)
 
+(** Common styles for GitHub TUI. *)
+module Style = Style
```
I'm no longer exporting `Style` from both `Pretty.Doc` and `Pretty.Layout`. It was an unfortunate coincidence anyway.

Now `Style` should be used as `Pretty.Style`

chshersh commented on lib/pretty/style.ml:
```diff
@@ -1,15 +1,44 @@
-let pr_closed = ANSITerminal.[ Bold; red ]
-let pr_open = ANSITerminal.[ Bold; green ]
-let pr_merged = ANSITerminal.[ Bold; magenta ]
+module Color = Ansifmt.Color
+module Styling = Ansifmt.Styling
+
+type t = Styling.t
+
+let ( & ) = Styling.( & )
```
I reexport things from `Ansifmt` to follow something kinda Domain-Driven Design and provide a small local wrapper around `ansifmt` so other modules inside GitHub TUI shouldn't depend on Ansifmt directly, only on relevant parts.

chshersh commented on lib/pretty/style.ml:
```diff
@@ -1,15 +1,44 @@
+let none = Styling.none
+
+(* Repository name. *)
+let repo = Styling.(bold & fg Color.blue)
```
It would be nice to write just `Styling.(bold & fg blue)` without the `Color` but I don't really want to open `Color`.

Eh, fine for now, not a big deal.

qexat replied:
Makes sense, perhaps we should re-export some of the common colors in `Styling`

chshersh replied:
Actually, I take it back. I'm doing the same `Color.name` in my website written in Elm:

* https://github.com/chshersh/chshersh.github.io/blob/417529c96192109df6205f45d455687cf1ce5ac2/src/Page/Main.elm#L55

I guess, words like `yellow` or `magenta` have the only implicit meaning of color. But for more exotic colors like "suva grey" or "gainsboro" it's not that clear that it's a color. So having a `Color.` prefix is better.

I'll keep the prefix everywhere for now for consistency. Usually consistency is superior to typing fewer characters occasionally.

So I wouldn't push for reexporting colors. In GitHub TUI, I'll probably introduce some specific colors locally for different themes, so it's not a problem at all.

qexat replied:
Fair enough, I was a bit skeptical anyway

chshersh commented on lib/pretty/style.ml:
```diff
@@ -1,15 +1,44 @@
+let pr_open = Styling.(bold & fg Color.green)
+let pr_merged = Styling.(bold & fg Color.magenta)
+
+let apply_background_hex_color ~hex text =
+  let color = Color.make_rgb_hex hex in
+  match color with
+  | None -> (text, `Light)
+  | Some color ->
+      let foreground_suggestion = Color.best_for_contrast ~threshold:80 color in
+      let contents = Printf.sprintf " %s " text in
+      let text = Styling.(wrap ~contents (bg color)) in
+      (text, foreground_suggestion)
```
This function was so ugly and unsafe 😭 

I'm glad it's now much prettier and saner!

chshersh commented on test/test_pretty/test_pretty.ml:
```diff
@@ -1 +1 @@
-let tests = List.flatten [ Test_color.tests ]
+let tests = List.flatten []
```
I was testing only my own Color implementation but now I don't have tests anymore, lol.

They'll be added later, don't worry. There're lots of tests to write.

chshersh commented:
@qexat This is mostly for you to review. Feel free to share your feedback or any comments 🙂

qexat reviewed (approved):
Looks good to me. `dune build @install` builds fine locally.

qexat committed:
Make Ansifmt a normal dependency
Ansifmt 0.2.0 was just released, so we can pull it from the opam repository directly.

qexat commented:
I made Ansifmt pulling its source from the Opam repository instead of from source, as 0.2.0 was just released: https://opam.ocaml.org/packages/ansifmt/

(The force-push was due to an incorrect formatting of the commit message :sweat_smile: I thought it would handle `\n` as a newline)

chshersh commented:
> I made Ansifmt pulling its source from the Opam repository instead of from source, as 0.2.0 was just released: https://opam.ocaml.org/packages/ansifmt/
> 
> (The force-push was due to an incorrect formatting of the commit message 😅 I thought it would handle `\n` as a newline)

Amazing! Thank you! 🙏🏻

heathhenley reviewed (approved):
🚀

heathhenley commented:
`dune build @install` is failing for me but I'm probably doing something noobish - do you have do something else to sync your env? 

![image](https://github.com/user-attachments/assets/1ea9d32b-33dd-4848-9fe6-8b6563c75f03)

qexat commented:
@heathhenley it seems that you don't have ansifmt installed in your current switch

try running:

```sh
opam install . --deps-only --with-test
```

heathhenley commented:
> @heathhenley it seems that you don't have ansifmt installed in your current switch
> 
> try running:
> 
> ```shell
> opam install . --deps-only --with-test
> ```

🤦- thanks all good now!

Does anyone think we should throw a note in the readme about how to update switch when deps change / update?  

I was also thinking about the list of 'non-obvious' things you mentioned in the other thread - wonder if using the log module to log debug info to a file would count (for dev)? And getting / setting your github token (for everyone)

chshersh commented:
> > @heathhenley it seems that you don't have ansifmt installed in your current switch
> > try running:
> > ```shell
> > opam install . --deps-only --with-test
> > ```
> 
> 🤦- thanks all good now!
> 
> Does anyone think we should throw a note in the readme about how to update switch when deps change / update?
> 
> I was also thinking about the list of 'non-obvious' things you mentioned in the other thread - wonder if using the log module to log debug info to a file would count (for dev)? And getting / setting your github token (for everyone)

I don't mind the idea for having tips for developers. Currently, we have a section in README for that. So feel free to add more to it, like updating a dependency or using logs 🙂 

We don't have lots of docs yet, so it's fine to write everything in README for now. In the future, we'll probably have a separate `docs/` directory with all the docs.

chshersh commented:
Thanks @qexat and @heathhenley for the review! Merging 🙂
